src/hclust.jl

```diff
@@ -531,6 +531,101 @@ function rorder!(hmer::HclustMerges)
     return hmer
 end
 
+
+"""
+    orderleaves_barjoseph!!(order::Vector{Int}, hcl::Hclust, dm::Array{Float64,2})
+
+Given a hierarchical cluster and the distance matrix used to generate it,
+use fast algorithm to determine optimal leaf order minimizing the distance
+between adjacent leaves. This is done using a heuristic where,
+when combining multi-leaf sub branches, only the outermost leaves are
+compared (a maximum of 4 comparisons per intersection).
+
+Sub branches are flipped if necessary to minimize the distance between adjacent
+nodes, and then the combined branches are treated as a block for future
+comparisons.
+
+Based on:
+[Bar-Joseph et. al. "Fast optimal leaf ordering for hierarchical clustering." _Bioinformatics_. (2001)](https://doi.org/10.1093/bioinformatics/17.suppl_1.S22)
+"""
+function orderleaves_barjoseph!(hcl::Hclust, dm::Array{Float64,2})
+    extents = Tuple{Int,Int}[]
+    for v in axes(hcl.merges, 1)
+        vl, vr = hcl.merges[v, 1], hcl.merges[v, 2]
+
+        (u, m, uidx, midx) = leaflocs(vl, hcl.order, extents)
+        (k, w, kidx, widx) = leaflocs(vr, hcl.order, extents)
+        if vl < 0 && vr < 0
+            # Nothing needs to be done
+        elseif vl < 0
+            # check if flipping would reduce distance
+            if dm[m,k] > dm[m,w]
+                reverse!(hcl.order, uidx, midx)
+                rotate_merges!(hcl.merges, vr)
+            end
+        elseif vr < 0
+            if dm[k,m] > dm[k,u]
+                reverse!(hcl.order, uidx, midx)
+                rotate_merges!(hcl.merges, vl)
+            end
+        elseif vl > 0 && vr > 0
+            # For 2 multi-leaf branches, determine if one or two flips is required
+            # 1 = do not flip
+            # 2 = flip left
+            # 3 = flip right
+            # 4 = flip both
+            flp = argmin((dm[m,k], dm[u,k], dm[m,w], dm[u,w]))
+            if flp == 2 || flp == 4
+                reverse!(hcl.order, uidx, midx)
+                rotate_merges!(hcl.merges, vr)
+            end
+            if flp == 3 || flp == 4
+                reverse!(hcl.order, kidx, widx)
+                rotate_merges!(hcl.merges, vl)
+            end
+        else
+            error("invalid 'merge' order in Hclust: ($vl, $vr) ")
+        end
+
+        push!(extents, (uidx, widx))
+    end
+end
+
+"""
+    leaflocs(v::Int, order::Vector{Int}, extents::Vector{Tuple{Int,Int}})
+
+`v`: vertex - may be a leaf (negative numbers) or a merge index
+`order`: Vector of leaf positions, same as hclust.order
+`extents`: tuples of the order indices for outermost leaves for each merge
+
+Returns the `order` values and indices of the extents for a given `v`.
+If `v` is a leaf, left and right extents will be the same.
+"""
+function leaflocs(v::Int, order::Vector{Int}, extents::Vector{Tuple{Int,Int}})
```

Comment:
Maybe rename it into something like (along with simplifying the returned values):
```suggestion
function leaves_range(v::Int, order::Vector{Int}, node_ranges::Vector{Tuple{Int,Int}})
```